api/v1/collection.go

```diff
@@ -1,69 +1,80 @@
 package api
 
 import (
+	"encoding/json"
 	"net/http"
 
+	"github.com/data-preservation-programs/spade-tenant/db"
 	"github.com/google/uuid"
+	"github.com/jackc/pgtype"
 	"github.com/labstack/echo/v4"
 )
 
-type MutableCollection struct {
-	Name                   string                  `json:"name"`
-	ReplicationConstraints []ReplicationConstraint `json:"replication_constraints"`
-	PieceListSource        PieceListSource         `json:"piece_list_source"`
-	Inactive               bool                    `json:"inactive"`
-	DealParams             DealParams              `json:"deal_params"`
+type CollectionResponse struct {
+	CollectionID          uuid.UUID    `json:"collection_id"`
+	CollectionName        string       `json:"name"`
+	CollectionActive      *bool        `json:"active"`
+	CollectionPieceSource pgtype.JSONB `json:"piece_list_source"`
+	CollectionDealParams  pgtype.JSONB `json:"deal_params"`
 }
 
-type DealParams struct {
-	DurationDays     uint `json:"duration_days"`
-	StartWithinHours uint `json:"start_within_hours"`
+func (a *apiV1) ConfigureCollectionRouter(e *echo.Group) {
+	g := e.Group("/collections")
+	g.GET("", a.handleGetCollections)
+	g.PUT("/:collectionUUID", a.handleModifyCollection)
+	g.POST("", a.handleCreateCollection)
+	g.DELETE("/:collectionUUID", a.handleDeleteCollection)
 }
 
-type Collection struct {
-	UUID uuid.UUID `json:"uuid"`
-	MutableCollection
-}
+// handleCreateCollection godoc
+//
+//	@Summary		Creates a new collection
+//	@Security		apiKey
+//	@Param 			collection CollectionMutable body true "Collection to create"
+//	@Produce		json
+//	@Success		200	{object}	ResponseEnvelope{response=Collection}
+//	@Router			/collections [post]
+func (a *apiV1) handleCreateCollection(c echo.Context) error {
+	var collection db.Collection
 
-type PieceListSource struct {
-	Method            string `json:"method"`
-	PollIntervalHours int    `json:"poll_interval_hours,omitempty"`
-	ConnectionDetails string `json:"connection_details"` // TODO: better types / validation for the connection details
-}
+	err := json.NewDecoder(c.Request().Body).Decode(&collection)
+	if err != nil {
+		return c.JSON(http.StatusInternalServerError, CreateErrorResponseEnvelope(c, http.StatusInternalServerError, err.Error()))
+	}
 
-type ReplicationConstraint struct {
-	ConstraintID  int `json:"constraint_id"`
-	ConstraintMax int `json:"constraint_max"`
-}
+	collection.TenantID = db.ID(GetTenantContext(c).TenantID)
+	collection.CollectionID = uuid.New()
```

Comment:
I think Gorm should automatically do this for you, so you don't need to specify it, right?

Reply:
Doesnt seem to. Even changing the schema to have default:uuid_generate_v4() doesnt seem to work. 

```ERROR: null value in column "collection_id" of relation "collections" violates not-null constraint (SQLSTATE 23502)
2024/01/03 11:09:42 stdout: [1.296ms] [rows:0] INSERT INTO "collections" ("updated_at","deleted_at","tenant_id","collection_name","collection_active","collection_piece_source","collection_deal_params") VALUES ('2024-01-03 11:09:42.776',NULL,'1','ppppp',true,'{"method":"method","poll_interval_hours":10,"connection_details":"details"}','{"duration_days":10,"start_within_hours":10}') RETURNING "created_at","collection_id","collection_piece_source","collection_deal_params"
2024/01/03 11:09:44 stdout: ```